### PR TITLE
Don't write zero-feature Analysis when retrieval crashes; guard blank middle names (prod port)

### PR DESCRIPTION
**Pull request by paulalbert1**
Clean cherry-pick of #712 (merged to `development` as `f19db941`) onto the Corretto17 prod line — zero conflicts, same 4 files.

## Problem

Identities with an empty-string `middleName` and a compound surname (e.g. `mah4006`, Manuel Hidalgo Medina) crash the retrieval worker in `deriveAdditionalName` → `AuthorName.substring(0,1)`. `ForkJoinPool.execute` swallows the exception entirely (no log, no stderr), so the request thread resumes with a null `ESearchResult`, scores zero candidates, and **overwrites the Analysis row with zeros as an HTTP 200** — the May 3–4 corruption shape, repeating on every retry since `ESearchResult` is never populated.

## Fix (three layers)

1. `deriveAdditionalName` treats a blank middle name as null (the crash site).
2. Retrieval workers record any `Throwable` (per-call concurrent set; Errors rethrown after recording); `retrieveArticlesByDateRange` returns false when a worker failed — its boolean was write-only before, verified across all callers.
3. A failed retrieval returns 500 from `retrieveArticlesByUid` and 404 from the feature-generator **with the prior Analysis row untouched**. Gate is strictly "retrieval failed", never "ESearchResult still null", so genuinely zero-article people keep today's behavior exactly.

## ⚠️ Deliberate behavior change

A transient retrieval `IOException` (PubMed 429 pressure) now yields 500/404 with the prior Analysis preserved, instead of 200 re-scoring stale candidates — expect nightly inst-client failure counts to rise visibly during PubMed flakiness. By design, consistent with #691.

## Verification

- `mvn test` on this branch: **238/238 pass**.
- Dev-verified after #712 deployed to reciter-dev (image `634.…f19db941`): `mah4006` went from a 0.16s zero-feature 200 to a real 58s retrieval — **411 features, recall 1.0** (all 353 accepted PMIDs retrieved); a normal uid regression check was byte-identical before/after.
- Crash reproduced standalone against the released identity-model 3.0.2 jar; companion root fix ReCiter-Identity-Model #14 (v3.0.3) is open separately and is not a blocker.

Replaces #713 (dev→master promotion), closed in favor of this targeted port to the line prod actually runs.

### CODE REVIEW

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java [hunk 1]
```diff
@@ -1016,30 +1035,29 @@ private void identityAuthorNames(Identity identity, Map<IdentityNameType, Set<Au
...
+
 		Set<AuthorName> derivedAuthorNames = new HashSet<AuthorName>();
+		// AuthorName's constructor takes substring(0, 1) of any non-null middleName, so a
+		// blank one (an Identity primaryName can carry middleName "") must behave exactly
+		// like null — both as a constructor argument and as a first-name substitute below.
+		String middleName = identityName.getMiddleName();
+		if(middleName != null && middleName.trim().isEmpty()) {
+			middleName = null;
+		}
 		if(identityName.getLastName().contains(" ") || identityName.getLastName().contains(".")) {
```
Comment (mrj4001) [inline src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Although getLastName() is expected to be non-null, it may be safer not to assume that the value is always present. Could we add a null/blank check before calling contains() and split()? This would prevent a potential NullPointerException if getLastName() unexpectedly returns null.

Suggested Fix:
String lastName = identityName.getLastName();

if (lastName != null && !lastName.trim().isEmpty()
        && (lastName.contains(" ") || lastName.contains("."))) {

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java [hunk 2]
```diff
@@ -1016,30 +1035,29 @@ private void identityAuthorNames(Identity identity, Map<IdentityNameType, Set<Au
...
 			String[] possibleLastName = identityName.getLastName().split("\\s+|-", 2);
```
Comment (mrj4001) [inline src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: It splits on whitespace or -, not a period. But not sure why are we checking  identityName.getLastName().contains(".") in if condition.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java [hunk 3]
```diff
@@ -1016,30 +1035,29 @@ private void identityAuthorNames(Identity identity, Map<IdentityNameType, Set<Au
...
-			if(possibleLastName[0].length() >=4 
+			if(possibleLastName[0].length() >=4
```
Comment (mrj4001) [inline src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Please verify that possibleLastName contains at least one element before accessing possibleLastName[0]. Although the current condition appears to prevent an empty last name from reaching this block, explicitly validating the array length would make the code safer and avoid a potential ArrayIndexOutOfBoundsException.
suggested Fix:
if (possibleLastName.length > 0 && possibleLastName[0].length() >= 4)

File: src/test/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngineTest.java
Comment (mrj4001) [inline src/test/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngineTest.java:1]: Should we add test cases to cover scenarios where the last name is an empty string ("") or contains a period (.)? This would help ensure these input conditions are properly handled and provide coverage for the corresponding code paths.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java [hunk 4]
```diff
@@ -122,17 +125,29 @@ public void run() {
 						+ startDate + "] endDate=[" + endDate + "].");
 					retrieveDataByDateRange(identity, startDate, endDate, this.refreshFlag);
 				}
-			} catch (IOException e) {
-				slf4jLogger.error("Unabled to retrieve. " + identity.getUid(), e);
+			} catch (Throwable t) {
```
Comment (mrj4001) [inline src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java:128]: The existing tests cover an unchecked exception from the retrieval worker and the no-work scenario. Could we also add a test that simulates an Error from the worker and verifies that the UID is recorded as failed before the Error is rethrown? This would directly validate the behavior described in the catch (Throwable) comment.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java [hunk 5]
```diff
@@ -1016,30 +1035,29 @@ private void identityAuthorNames(Identity identity, Map<IdentityNameType, Set<Au
...
-					middleName = identityName.getMiddleName();
-				}
+
 				AuthorName authorName1 = new AuthorName(identityName.getFirstName(), middleName, possibleLastName[0].trim());
 				AuthorName authorName2 = new AuthorName(identityName.getFirstName(), middleName, possibleLastName[1].trim());
 				derivedAuthorNames.add(authorName1);
 				derivedAuthorNames.add(authorName2);
 			}
 		}
 		if(identityName.getFirstName().contains(" ") || identityName.getFirstName().contains(".")) {
```
Comment (mrj4001) [inline src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Could we validate/normalize firstName before checking its length? Currently, length() == 2 is evaluated on the untrimmed value, so a value such as "W. " would not satisfy the condition even though trim().endsWith(".") is true. Also, please consider handling a possible null value for getFirstName() before calling contains().

Comment (paulalbert1) [reply src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Some users genuinely have a single letter as a first name like one of our chairs

File: src/main/java/reciter/controller/ReCiterController.java [hunk 6]
```diff
@@ -1285,15 +1295,28 @@ private EngineParameters initializeEngineParameters(String uid, Double totalStan
...
+                    // Analysis row with a zero-feature result. Bail out to the caller's 404.
+                    log.error("Retrieval failed for uid={} (status={}); skipping feature generation",
+                            uid, retrievalResponse.getStatusCode());
+                    return null;
+                }
                 eSearchResults = eSearchResultService.findByUid(uid);
             } else if(eSearchResults != null && (retrievalRefreshFlag == RetrievalRefreshFlag.ALL_PUBLICATIONS || retrievalRefreshFlag == RetrievalRefreshFlag.ONLY_NEWLY_ADDED_PUBLICATIONS)) {
-            	retrieveArticlesByUid(uid, retrievalRefreshFlag, fullSweepMaxAgeDays);
+            	ResponseEntity retrievalResponse = retrieveArticlesByUid(uid, retrievalRefreshFlag, fullSweepMaxAgeDays);
+            	if (retrievalResponse.getStatusCode().isError()) {
```
Comment (mrj4001) [inline src/main/java/reciter/controller/ReCiterController.java]: Please check for retrievalResponse is not null prior to calling getStatusCode()

File: src/main/java/reciter/controller/ReCiterController.java [hunk 7]
```diff
@@ -1285,15 +1295,28 @@ private EngineParameters initializeEngineParameters(String uid, Double totalStan
...
                 log.warn("No ESearchResult for uid={}; upgrading requested flag={} to ALL_PUBLICATIONS",
                          uid, retrievalRefreshFlag);
-                retrieveArticlesByUid(uid, RetrievalRefreshFlag.ALL_PUBLICATIONS);
+                ResponseEntity retrievalResponse = retrieveArticlesByUid(uid, RetrievalRefreshFlag.ALL_PUBLICATIONS);
                 // Marked AFTER the call: retrieveArticlesByUid resets the tracker on entry.
                 // Auto-upgrades are reported separately from escalations (#696) — they are
                 // unbudgeted full sweeps, and folding them into the escalation count would
                 // make the client's "budget is holding" signal a false one.
                 RetrievalEscalationTracker.markAutoUpgraded();
+                if (retrievalResponse.getStatusCode().isError()) {
```
Comment (mrj4001) [inline src/main/java/reciter/controller/ReCiterController.java]: Please check for retrievalResponse is not null prior to calling getStatusCode()

**Review by mrj4001 — CHANGES_REQUESTED** [review]
Please address the code review comments.

**Comment by paulalbert1** [comment]
Reviewed this alongside #716, with the findings independently re-verified against the code rather than accepted on first pass. **Recommend merging, and merging this one first** — #716 is stacked on it. No blocking findings.

## Port fidelity — verified mechanically

The "clean cherry-pick" claim holds byte-for-byte, checked two independent ways:

- **Diff-of-diffs.** `git diff 8136a8b3 581e92df` (upstream #712) versus `git diff origin/MigrationFromJDk11ToAWSCorretto17...origin/fix/empty-middlename-corretto17` are byte-identical — 409 lines, `diff` exit 0. No dropped hunks, altered logic, changed constants, lost null-checks, or extra riders.
- **Post-state comparison**, which is stronger since it also catches base drift: all four touched files hash-identical between `development` post-fix and the simulated merge of this PR onto the Corretto17 tip. Every upstream test is present; none weakened.

## Target-line safety

The two lines diverge far less than the Corretto17 label suggests — only four main-source files, none of them on this PR's path. Notably `pom.xml` is **identical** between `development` and `MigrationFromJDk11ToAWSCorretto17` (reciter-identity-model 3.0.2 on both), so the guard's assumption about `AuthorName`'s constructor rests on the same artifact on both lines. `ExternalArticleController` never calls retrieval, feature generation, or `ArticleTranslator`, so the new gate neither breaks nor bypasses the ExternalArticle path.

Caller hunt across the tree: `retrieveArticlesByDateRange` has one production implementation and three call sites — `ReCiterController:396` (the all-identities endpoint, which ignores the boolean on `development` identically, so not a port-introduced divergence) plus the two gated sites at `:457` and `:530`. `retrieveArticlesByUid` has one internal caller and one public overload, both accounted for. `initializeEngineParameters` has a single caller at `:885` that already null-checks into 404.

## Verification run

| Check | Result |
|---|---|
| `mvn test` at this tip (JDK 17) | **238 tests, 0 failures, 0 errors, 0 skipped** — matches the PR body exactly |
| Merged tree (this PR + #716 onto Corretto17 tip), built and run | **258 tests, 0 failures**, ported tests green alongside `ExternalArticleControllerFeedbackTest`, `DynamoDbGoldStandardServiceRaceTest`, `DynamoDbGoldStandardServiceFeedbackDiffTest` |
| `git merge-tree --write-tree` | Conflict-free |

## Concurrency and gate logic — checked specifically, clean

- `failedUids` is a method-local `ConcurrentHashMap.newKeySet()` created per `retrieveArticlesByDateRange` invocation and passed by reference into each worker. It is per-call and thread-safe; two concurrent requests for different uids cannot observe each other's failure state. No static or shared mutable failure flag was introduced.
- The `Error` rethrow is correctly ordered — log, then `failedUids.add`, then `throw` — so the set is consistent before the rethrow, and the rethrow only completes the `ForkJoinTask` exceptionally without disturbing `shutdown()`/`awaitTermination()`. The gate still fires.
- The rethrow is **not** dead code: on JDK 17, `Executors.newWorkStealingPool(15).execute(...)` returns true from `awaitTermination` even when a worker dies, so the boolean gate is genuinely the only way the caller learns of a crash. (The inline comment claiming the pool swallows the rethrow is inaccurate, but the code has real effect.)
- A legitimately empty result is **not** misclassified — `AsyncRetrievalEngine.run` records a uid only when a `Throwable` escapes, so a person with genuinely zero articles still gets today's clean 200.
- Mutation-tested: reverting the blank-middleName normalization fails 3 tests; reverting the `failedUids`/catch-`Throwable` recording fails 1; reverting the two HTTP-500 gates fails 2. The assertions are real.

## Two non-blocking notes

**1. Half the corruption path remains open — filed as #720.** The gate fires only when a worker actually *throws*. The two dominant real-world failures don't: `AbstractRetrievalStrategy:283-291` and `PubMedArticleRetriever:52-60` catch the exception, call `RetrievalErrorTracker.markError()`, and return empty. That tracker is a worker-thread `ThreadLocal` the request-thread gate never consults. On the `ALL_PUBLICATIONS` path — which deletes the ESearchResult at `:448` first — a PubMed-tool outage still ends at `analysisService.save` with `reCiterArticleFeatures: 0` and HTTP 200. This is pre-existing, and this PR strictly *reduces* exposure, so it is a follow-up rather than a change request. Full trace and three candidate fixes in #720.

**2. The second gate is untested.** Deleting `ReCiterController.java:1316-1320` (the `else if` refresh-with-existing-ESearchResult branch) leaves all 61 `ReCiterControllerTest` tests green. As a control, deleting the *first* gate is killed immediately by `testRunFeatureGeneratorAutoUpgradeRetrievalCrashReturns404AndNeverSavesAnalysis`. All nine `runFeatureGenerator` calls in the test pass `null` for the refresh flag, so the branch is statically unreachable from the suite — and it is the branch the nightly institutional client actually drives. Worth one test before or shortly after merge.

## Two housekeeping items

- This branch is based on `c6783613`, 13 commits behind the Corretto17 tip. Harmless — `git diff --stat c6783613 f5142347` over the four ported files is empty — but no CI checks are configured on the PR, so nothing is exercising the current tip automatically. That's why I built and tested the merge result directly.
- The deliberate behavior change in the PR body is worth restating for whoever watches the nightly: transient PubMed `IOException`s now surface as 500/404 instead of a 200 over stale candidates, so failure counts will rise visibly during PubMed flakiness. That is the point of the change, not a symptom.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java (hunk 1 above)
Comment (paulalbert1) [reply src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Done in b7dabdc. Hoisted the value and guarded it:

```java
String lastName = identityName.getLastName();
if (lastName != null && !lastName.isBlank()
        && (lastName.contains(" ") || lastName.contains("."))) {
```

Used isBlank() rather than trim().isEmpty() to match your preference on the other thread. Covered by blankAndNullLastNamesDeriveNothingInsteadOfThrowing, which builds the null case via the no-arg AuthorName constructor (the setters reject null, but DynamoDB deserialization leaves the fields null, which is how this reaches us).

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java (hunk 2 above)
Comment (paulalbert1) [reply src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Good catch, and it was worse than an inconsistency: it was a live crash.

The condition admits a period, but the pattern splits on whitespace and hyphens only, so a period-without-space surname enters the branch and splits into ONE element. Verified:

```
lastName="St.John"  entersBranch=true  splitLen=1  -> ArrayIndexOutOfBoundsException
lastName="O.Brien"  entersBranch=true  splitLen=1  -> ArrayIndexOutOfBoundsException
lastName="St. John" entersBranch=true  splitLen=2  -> ok
```

That throws inside the retrieval worker, which is exactly the swallowed-crash path this PR exists to close, so it would have produced the same zero-feature Analysis row with a 200.

I fixed it with the length guard rather than by adding "." to the split pattern, because splitting on periods would create derived names we do not generate today and that is a retrieval-behaviour change, not a bug fix. If you would rather "St.John" actually derive St / John, say so and I will do it as a separate change with its own test.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java (hunk 3 above)
Comment (paulalbert1) [reply src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Done in b7dabdc, with one adjustment: line 1052 reads possibleLastName[1], so length > 0 still throws. Used length > 1:

```java
if (possibleLastName.length > 1
        && possibleLastName[0].length() >= 4
        && possibleLastName[1].length() >= 4) {
```

Mutation-checked: replacing that condition with `true` fails periodWithoutSpaceSurnameDerivesNothingInsteadOfThrowing with `ArrayIndexOutOfBounds Index 1 out of bounds for length 1`.

File: src/test/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngineTest.java
Comment (paulalbert1) [reply src/test/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngineTest.java:1]: Added both, plus a regression guard so the fix does not silently disable the feature:

- periodWithoutSpaceSurnameDerivesNothingInsteadOfThrowing ("St.John", "O.Brien")
- periodWithSpaceSurnameStillDerivesBothHalves ("Della. Robbia" still splits)
- blankAndNullLastNamesDeriveNothingInsteadOfThrowing ("", " ", null)
- nullFirstNameDerivesWithoutThrowing

AliasReCiterRetrievalEngineTest goes from 6 to 13 tests; full suite 245, 0 failures.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java (hunk 4 above)
Comment (paulalbert1) [reply src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java:128]: Added errorFromRetrievalWorkerIsRecordedBeforeItIsRethrown. A test subclass overrides retrieveData to throw StackOverflowError; retrieveArticlesByDateRange returning false is the proof the uid reached failedUids before the rethrow left the catch block.

This needed one seam: retrieveData is now package-private rather than private so the override is possible. Flagging it explicitly since you asked about a widening on #716 — same rationale, and the file already uses that pattern for deriveAdditionalName. Happy to revert it and drop the test if you would rather not widen it.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java (hunk 5 above)
Comment (paulalbert1) [reply src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java]: Both applied in b7dabdc:

```java
String firstName = identityName.getFirstName();
if (firstName != null && (firstName.contains(" ") || firstName.contains("."))) {
    String trimmedFirstName = firstName.trim();
    if (trimmedFirstName.length() == 2 && trimmedFirstName.endsWith(".") && middleName != null) {
```

On Paul's point about genuine one-letter first names: I traced it and the trim does not affect them. A real one-letter first name ("W") contains neither a space nor a period, so it never enters this branch at all; it derives via the length()==1 case further down, unchanged. The trim only changes values that already end in "." once trimmed, i.e. "W. " now behaves like "W.". Pinned by genuineOneLetterFirstNameIsUnaffectedByTheTrimNormalization so it cannot regress silently.

Also added the null guard to the length()==1 case below, which had the same exposure.

File: src/main/java/reciter/controller/ReCiterController.java (hunk 6 above)
Comment (paulalbert1) [reply src/main/java/reciter/controller/ReCiterController.java]: Done in b7dabdc: `if (retrievalResponse == null || retrievalResponse.getStatusCode().isError())`.

File: src/main/java/reciter/controller/ReCiterController.java (hunk 7 above)
Comment (paulalbert1) [reply src/main/java/reciter/controller/ReCiterController.java]: Done in b7dabdc, same change at this gate: `if (retrievalResponse == null || retrievalResponse.getStatusCode().isError())`.

File: src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java [hunk 8]
```diff
@@ -1045,36 +1046,51 @@ Set<AuthorName> deriveAdditionalName(AuthorName identityName) {
...
 		}
-		if(identityName.getLastName().contains(" ") || identityName.getLastName().contains(".")) {
-			String[] possibleLastName = identityName.getLastName().split("\\s+|-", 2);
-			if(possibleLastName[0].length() >=4
+		// A malformed Identity can carry a null first or last name (see #715/#717), and an
+		// unchecked throw here dies inside a retrieval worker — the swallowed-crash shape
+		// this PR exists to stop. Read each once, guard each once.
+		String lastName = identityName.getLastName();
+		String firstName = identityName.getFirstName();
+		if(lastName != null && !lastName.trim().isEmpty()
```
Comment (mrj4001) [inline src/main/java/reciter/xml/retriever/engine/AliasReCiterRetrievalEngine.java:1054]: Please conside isBlank instead of isEmpty()
